src/mc-writer.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2022-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongocrypt-private.h"
+
+#include "mc-writer-private.h"
+
+#define CHECK_AND_RETURN(x) \
+   if (!(x)) {              \
+      return false;         \
+   }
+
+#define CHECK_REMAINING_BUFFER_AND_RET(write_size)       \
+   if ((writer->pos + (write_size)) > writer->len) {     \
```

Comment:
```suggestion
   if ((write_size) > writer->len - writer->pos) {     \
```
Avoid potential unsigned integer overflow.